src/main/scala/is/hail/variant/VariantDataset.scala

```diff
@@ -0,0 +1,1360 @@
+          (id, JSONAnnotationImpex.importAnnotation(jv, saSignature, "sample_annotations"))
+        case other => fatal(
+          s"""corrupt VDS: invalid metadata
+             |  Invalid sample annotation metadata
+             |  Recreate VDS with current version of Hail.""".stripMargin)
+      }
+      .toArray
+
+    val globalAnnotation = JSONAnnotationImpex.importAnnotation(getAndCastJSON[JValue]("global_annotation"),
+      globalSignature, "global")
+
+    val ids = sampleInfo.map(_._1)
+    val annotations = sampleInfo.map(_._2)
+
+    VariantMetadata(ids, annotations, globalAnnotation,
+      saSignature, vaSignature, globalSignature, wasSplit, isDosage)
+  }
+
+  def readKudu(hc: HailContext, dirname: String, tableName: String,
+    master: String): VariantDataset = {
+
+    val metadata = readMetadata(hc.hadoopConf, dirname, requireParquetSuccess = false)
+    val vaSignature = metadata.vaSignature
+    val isDosage = metadata.isDosage
+
+    val df = hc.sqlContext.read.options(
+      Map("kudu.table" -> tableName, "kudu.master" -> master)).kudu
+
+    val rowType = kuduRowType(vaSignature)
+    val schema: StructType = KuduAnnotationImpex.exportType(rowType).asInstanceOf[StructType]
+
+    // Kudu key fields are always first, so we have to reorder the fields we get back
+    // to be in the column order for the flattened schema *before* we unflatten
+    val indices: Array[Int] = schema.fields.zipWithIndex.map { case (field, rowIdx) =>
+      df.schema.fieldIndex(field.name)
+    }
+
+    val rdd: RDD[(Variant, (Annotation, Iterable[Genotype]))] = df.rdd.map { row =>
+      val importedRow = KuduAnnotationImpex.importAnnotation(
+        KuduAnnotationImpex.reorder(row, indices), rowType).asInstanceOf[Row]
+      val v = importedRow.getVariant(0)
+      (v,
+        (importedRow.get(1),
+          importedRow.getGenotypeStream(v, 2, metadata.isDosage)))
+    }.spanByKey().map(kv => {
+      // combine variant rows with different sample groups (no shuffle)
+      val variant = kv._1
+      val annotations = kv._2.head._1
+      // just use first annotation
+      val genotypes = kv._2.flatMap(_._2) // combine genotype streams
+      (variant, (annotations, genotypes))
+    })
+    new VariantSampleMatrix[Genotype](hc, metadata, rdd.toOrderedRDD)
+  }
+
+  def kuduRowType(vaSignature: Type): Type = TStruct("variant" -> Variant.t,
+    "annotations" -> vaSignature,
+    "gs" -> GenotypeStream.t,
+    "sample_group" -> TString)
+
+  private def makeSchemaForKudu(vaSignature: Type): StructType =
+    StructType(Array(
+      StructField("variant", Variant.schema, nullable = false),
+      StructField("annotations", vaSignature.schema, nullable = false),
+      StructField("gs", GenotypeStream.schema, nullable = false),
+      StructField("sample_group", StringType, nullable = false)
+    ))
+}
+
+class VariantDatasetFunctions(private val vds: VariantSampleMatrix[Genotype]) extends AnyVal {
```

Comment:
I was thinking of deleting the `VariantDataset` type alias and just make this `VariantDataset` and have the `HailContext` functions.  What's the barrier to doing this now?